File: crates/mithril-oxide-sys/cpp/IR/BuiltinAttributes.cpp
#include "BuiltinAttributes.hpp"

#include <memory>

#include <llvm/ADT/APSInt.h>
#include <llvm/ADT/StringRef.h>
#include <mlir/IR/BuiltinAttributes.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/Support/LLVM.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys {

const void* StringAttr_get(MLIRContext &context, rust::Str value)
{
    return StringAttr::get(
        &context,
        mlir::StringRef(value.data(), value.length())
    ).getAsOpaquePointer();
}

const void* FlatSymbolRefAttr_get(MLIRContext &context, rust::Str value)
{
    return FlatSymbolRefAttr::get(
        &context,
        mlir::StringRef(value.data(), value.length())
    ).getAsOpaquePointer();
}

const void* IntegerAttr_get(MLIRContext &context, rust::Str value)
{
    auto int_val = llvm::APSInt(llvm::StringRef(value.data(), value.length()));
    return IntegerAttr::get(&context, int_val).getAsOpaquePointer();
}

const void* BoolAttr_get(MLIRContext &context, bool value)
{
    return BoolAttr::get(&context, value).getAsOpaquePointer();
}

const void* DenseElementsAttr_get(
    const void* shaped_type,
    rust::Slice<const void *const> values
)
{
    std::vector<Attribute> values_vec;

    for (const auto &value : values)
        values_vec.push_back(Attribute::getFromOpaquePointer(value));

    return DenseElementsAttr::get(
        ShapedType::getFromOpaquePointer(shaped_type), values_vec
        )
        .getAsOpaquePointer();
}

const void* DenseFPElementsAttr_get(
    const void* shaped_type,
    rust::Slice<const void *const> values
)
{
    std::vector<Attribute> values_vec;

    for (const auto &value : values)
        values_vec.push_back(Attribute::getFromOpaquePointer(value));

    return DenseFPElementsAttr::get(
        ShapedType::getFromOpaquePointer(shaped_type), values_vec
        )
        .getAsOpaquePointer();
}

const void* DenseIntElementsAttr_get(
    const void* shaped_type,
    rust::Slice<const void *const> values
)
{
    std::vector<Attribute> values_vec;

    for (const auto &value : values)
        values_vec.push_back(Attribute::getFromOpaquePointer(value));

    return DenseIntElementsAttr::get(
        ShapedType::getFromOpaquePointer(shaped_type), values_vec
        )
        .getAsOpaquePointer();
}

const void* DenseI64ElementsAttr_get(
    const void* shaped_type,
    rust::Slice<const void *const> values
)
{
    std::vector<Attribute> values_vec;

    for (const auto &value : values)
        values_vec.push_back(Attribute::getFromOpaquePointer(value));

    return DenseIntElementsAttr::get(
        ShapedType::getFromOpaquePointer(shaped_type), values_vec
        )
        .getAsOpaquePointer();
}

const void* DictionaryAttr_get(
    MLIRContext &context,
    rust::Slice<const NamedAttribute *const> values
)
{
    std::vector<NamedAttribute> values_vec;

    for (const auto &value : values)
        values_vec.push_back(*value);

    return DictionaryAttr::get(&context, values_vec).getAsOpaquePointer();
}

const void* DenseBoolArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<bool> values
)
{
    return DenseBoolArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseI8ArrayAttr_get(
    MLIRContext &context,
    rust::Slice<rust::i8> values
)
{
    return DenseI8ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseI16ArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<rust::i16> values
)
{
    return DenseI16ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseI32ArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<rust::i32> values
)
{
    return DenseI32ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseI64ArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<rust::i64> values
)
{
    return DenseI64ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseF32ArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<rust::f32> values
)
{
    return DenseF32ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

const void* DenseF64ArrayAttr_get(
    MLIRContext &context,
    const rust::Slice<rust::f64> values
)
{
    return DenseF64ArrayAttr::get(&context, mlir::ArrayRef(values.data(), values.size())).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys

File: crates/mithril-oxide-sys/src/IR/BuiltinAttributes.rs
pub use self::ffi::{
    ArrayAttr_get, BoolAttr_get, DenseBoolArrayAttr_get, DenseElementsAttr_get,
    DenseF32ArrayAttr_get, DenseF64ArrayAttr_get, DenseFPElementsAttr_get, DenseI16ArrayAttr_get,
    DenseI32ArrayAttr_get, DenseI64ArrayAttr_get, DenseI8ArrayAttr_get, DenseIntElementsAttr_get,
    DictionaryAttr_get, FlatSymbolRefAttr_get, IntegerAttr_get, OpaqueAttr_get,
    SparseElementsAttr_get, StridedLayoutAttr_get, StringAttr_get, TypeAttr_get, UnitAttr_get,
};

#[cxx::bridge]
pub(crate) mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/BuiltinAttributes.hpp");

        type NamedAttribute = crate::IR::Attributes::NamedAttribute;
        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/BuiltinAttributes.hpp");

        type c_void = crate::c_void;

        #[must_use]
        fn UnitAttr_get(context: Pin<&mut MLIRContext>) -> *const c_void;
        #[must_use]
        unsafe fn TypeAttr_get(ttype: *const c_void) -> *const c_void;
        #[must_use]
        fn StringAttr_get(context: Pin<&mut MLIRContext>, value: &str) -> *const c_void;
        #[must_use]
        fn FlatSymbolRefAttr_get(context: Pin<&mut MLIRContext>, value: &str) -> *const c_void;
        #[must_use]
        fn IntegerAttr_get(context: Pin<&mut MLIRContext>, value: &str) -> *const c_void;
        #[must_use]
        fn BoolAttr_get(context: Pin<&mut MLIRContext>, value: bool) -> *const c_void;

        #[must_use]
        unsafe fn StridedLayoutAttr_get(
            context: Pin<&mut MLIRContext>,
            offset: i64,
            strides: &[i64],
        ) -> *const c_void;

        #[must_use]
        unsafe fn OpaqueAttr_get(
            dialect: *const c_void, // StringAttr
            attr_data: &str,
            ttype: *const c_void, // Type
        ) -> *const c_void;

        #[must_use]
        unsafe fn SparseElementsAttr_get(
            shaped_type: *const c_void, // ShapedType
            indices: *const c_void,     // DenseElementsAttr
            values: *const c_void,      // DenseElementsAttr
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseElementsAttr_get(
            shaped_type: *const c_void, // any type implementing ShapedType trait.
            // Attribute
            values: &[*const c_void],
        ) -> *const c_void;
        #[must_use]
        unsafe fn DenseFPElementsAttr_get(
            shaped_type: *const c_void, // any type implementing ShapedType trait.
            // Attribute
            values: &[*const c_void],
        ) -> *const c_void;
        #[must_use]
        unsafe fn DenseIntElementsAttr_get(
            shaped_type: *const c_void, // any type implementing ShapedType trait.
            // Attribute
            values: &[*const c_void],
        ) -> *const c_void;
        #[must_use]
        unsafe fn DictionaryAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[*const NamedAttribute],
        ) -> *const c_void;

        #[must_use]
        unsafe fn ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            // Attribute
            values: &[*const c_void],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseBoolArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[bool],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseI8ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[i8],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseI16ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[i16],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseI32ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[i32],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseI64ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[i64],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseF32ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[f32],
        ) -> *const c_void;

        #[must_use]
        unsafe fn DenseF64ArrayAttr_get(
            context: Pin<&mut MLIRContext>,
            values: &[f64],
        ) -> *const c_void;
    }
}

#[cfg(test)]
mod tests {
    use crate::IR::{Attributes::Attribute_print, MLIRContext::MLIRContext};

    use super::*;

    #[test]
    fn string_attr_new() {
        let mut context = MLIRContext::new();
        let string_attr = StringAttr_get(context.pin_mut(), "hello_world");
        assert_eq!(
            "\"hello_world\"",
            unsafe { Attribute_print(string_attr) }.as_str()
        );
    }
}
